### PR TITLE
Performance improvement and small fixes for flow issues in Mapgen::updateLiquid

### PR DESCRIPTION
**Pull request by MillersMan**
I've observed two issues with MapGen::updateLiquid that will be fixed with the request:

1. Water columns that end at the ground will not flow to the side when they are generated: http://imgur.com/aJNLdLC
2. Water columns that exceed the generated blocks will flow out even though there will be water added above when the upper blocks are generated: http://imgur.com/cHjqHYi

**Note**: this changes have to be applied to #4061 also.

### CODE REVIEW

**Comment by paramat** [comment]
These are fixes for water columns, which are not generated by current mapgens.
Instead, could you tune this PR to fix issues with current mapgens, i think the main one is water not flowing into an underwater tunnel entrance. Also in mgvalleys there may be some issues with the rivers.

**Comment by MillersMan** [comment]
It already fixes the underwater tunnel entrances. No further changes needed, ti also improves performance as it adds less nodes to the transformation queue than before.

**Comment by MillersMan** [comment]
The last commit prevents fluid layers that are only one node deep from being added the the queue twice.

**Comment by MillersMan** [comment]
I did tests with valley mapgen and the results so far are:
- Worst case: Shallow ocean ~100 transformations per block (would expect ~500 for the blocks theemselves, but average is over all blocks even those above and below)
- Surface and ground of deep oceans: below 100 per block (don't have exact values at it is spoiled with result from worst case scenario, would expect ~250 but as before)
- Valleys with rivers: Below 10 transformations per block (would expect 2 per column if column is higher than one node, otherwise it would be one, so here it is the same, blocks without liquids drop the rate)
- Deserts or other areas without liquids: No transformations

**Comment by MillersMan** [comment]
MapGen::updateLiquid statistics on my system:

- Before the PR:
 - time in ms: 11.9683 (min=10, max=24)
 - trans_liquid adds: 16076.1 (min=12544, max=34048)

- As currently in the PR:
 - time in ms: 10.5825 (min=8, max=26)
 - trans_liquid adds: 2930.07 (min=0, max=21487)

- With checking neighbours (not pushed yet):
 - time in ms: 32.4406 (min=8, max=151)
 - trans_liquid adds: 111.995 (min=0, max=1330)

**Comment by paramat** [comment]
Interesting, this work is much appreciated.

**Comment by MillersMan** [comment]
New version with all changes squashed and some further improvements:

- Added back a few missing adds to trans_liquid, so it is currently at arround 170 per chunk over a large area of the map
- Needs only around 10.5 ms on my system compared to around 12 ms of the old version

I think this version is ready to be merged.

**Comment by Fixer-007** [comment]
Will play with it :+1: Including 4061 in compilation.

**Comment by sofar** [comment]
- space after `if` and before `(` please

if there's only one user of the helper function, isn't it perhaps a good idea to make it a static inline instead, and not add it to the class?

**Comment by MillersMan** [comment]
I've inserted the space between `if` and `(` and also changed the helper to be a static inline function

File: src/mapgen.cpp [hunk 1]
```diff
@@ -199,27 +199,89 @@ void Mapgen::updateHeightmap(v3s16 nmin, v3s16 nmax)
 	//printf("updateHeightmap: %dus\n", t.stop());
 }
 
+inline static bool updateLiquidHelper(Mapgen *mg, int i, const v3s16 &em)
```
Comment (nerzhul) [inline src/mapgen.cpp]: Please make this function member of Mapgen

File: src/mapgen.cpp [hunk 2]
```diff
@@ -199,27 +199,86 @@ void Mapgen::updateHeightmap(v3s16 nmin, v3s16 nmax)
 	//printf("updateHeightmap: %dus\n", t.stop());
 }
 
+inline bool Mapgen::updateLiquidHelper(int i, const v3s16 &em)
```
Comment (kwolekr) [inline src/mapgen.cpp]: Hrmm?  What does this function 'help' with?
What is `i` here?
What is `em`?
Why bother passing a const ref to em?  It's the size of a machine word on the most common platform Minetest will be run on.

File: src/mapgen.h [hunk 3]
```diff
@@ -177,6 +177,7 @@ class Mapgen {
 	s16 findGroundLevel(v2s16 p2d, s16 ymin, s16 ymax);
 	s16 findLiquidSurface(v2s16 p2d, s16 ymin, s16 ymax);
 	void updateHeightmap(v3s16 nmin, v3s16 nmax);
+	bool updateLiquidHelper(int i, const v3s16 &em);
```
Comment (kwolekr) [inline src/mapgen.h]: Probably should be `private`, and should share the `inline` decorator here as well.

File: src/mapgen.cpp [hunk 4]
```diff
@@ -199,27 +199,86 @@ void Mapgen::updateHeightmap(v3s16 nmin, v3s16 nmax)
...
-				// there was a change between liquid and nonliquid, add to queue.
-				if (isliquid != wasliquid)
+	for (s16 z = nmin.Z + 1; z <= nmax.Z - 1; z++)
+	for (s16 x = nmin.X + 1; x <= nmax.X - 1; x++) {
+		wasignored = true;
+		wasliquid = false;
+		waschecked = false;
+		waspushed = false;
+
+		u32 i = vm->m_area.index(x, nmax.Y, z);
```
Comment (kwolekr) [inline src/mapgen.cpp]: For what it's worth, the convention is to use `vi` as the name of a variable that represents the index of a node inside a VoxelManip.

File: src/mapgen.cpp [hunk 5]
```diff
@@ -199,27 +199,86 @@ void Mapgen::updateHeightmap(v3s16 nmin, v3s16 nmax)
...
+				bool ispushed = false;
+				if (updateLiquidHelper(i, em)) {
 					trans_liquid->push_back(v3s16(x, y, z));
-
-				wasliquid = isliquid;
-				vm->m_area.add_y(em, i, -1);
+					ispushed = true;
+				}
+				// Remember waschecked and waspushed to avoid repeated 
+				// checks/pushes in case the column consists of only this node
```
Comment (kwolekr) [inline src/mapgen.cpp]: Neat optimization here!

File: src/mapgen.cpp [hunk 6]
```diff
@@ -199,27 +199,86 @@ void Mapgen::updateHeightmap(v3s16 nmin, v3s16 nmax)
...
+				waschecked = true;
+				waspushed = ispushed;
+			} else {
+				// This is the topmost node below a liquid column
+				u32 j = i;
+				vm->m_area.add_y(em, j, 1);
+				if (!waspushed && (ndef->get(vm->m_data[i]).floodable ||
```
Comment (kwolekr) [inline src/mapgen.cpp]: Correct me if I'm wrong but:
The code block this if statement is in only gets executed when there is a transition from a liquid to a non liquid.
This piece of logic checks if the node *above* the transition node is floodable - but we already said the above node is always a liquid.  Liquids are not typically floodable.  Does that mean this ndef->get(vm->m_data[i].floodable condition will almost always be true?  Is it necessary then, if that's the case?

File: src/mapgen.cpp [hunk 7]
```diff
@@ -199,27 +199,86 @@ void Mapgen::updateHeightmap(v3s16 nmin, v3s16 nmax)
 	//printf("updateHeightmap: %dus\n", t.stop());
 }
 
+inline bool Mapgen::checkSidewardFloodable(int vi, v3s16 em)
```
Comment (kwolekr) [inline src/mapgen.cpp]: vi should be u32 here

File: src/mapgen.h [hunk 8]
```diff
@@ -197,6 +197,7 @@ class Mapgen {
 	virtual int getSpawnLevelAtPoint(v2s16 p) { return 0; }
 
 private:
+	inline bool checkSidewardFloodable(int vi, v3s16 em);
```
Comment (kwolekr) [inline src/mapgen.h]: Should add a comment indicating this function is a helper for updateLiquids()

File: src/mapgen.cpp (hunk 7 above)
Comment (kwolekr) [inline src/mapgen.cpp]: Err.. don't you mean Sideways??

**Pull request merged**